pkg/cosign/tuf/client.go

```diff
@@ -235,66 +233,62 @@ func GetRootStatus(ctx context.Context) (*RootStatus, error) {
 //       targets in a targets/ subfolder.
 //   * forceUpdate: indicates checking the remote for an update, even when the local
 //       timestamp.json is up to date.
-func initializeTUF(ctx context.Context, mirror string, root []byte, embedded fs.FS, forceUpdate bool) (*TUF, error) {
-	singletonTUFMu.Lock()
-	defer singletonTUFMu.Unlock()
-	if singletonTUF != nil {
-		return singletonTUF, nil
-	}
+func initializeTUF(ctx context.Context, mirror string, root []byte, embedded fs.FS, forceUpdate bool) *TUF {
+	singletonTUFOnce.Do(func() {
+		t := &TUF{
+			mirror:   mirror,
+			embedded: embedded,
+		}
 
-	t := &TUF{
-		mirror:   mirror,
-		embedded: embedded,
-	}
+		t.targets = newFileImpl()
+		var err error
+		t.local, err = newLocalStore()
+		if err != nil {
+			panic(err)
+		}
 
-	t.targets = newFileImpl()
-	var err error
-	t.local, err = newLocalStore()
-	if err != nil {
-		return nil, err
-	}
+		t.remote, err = remoteFromMirror(ctx, t.mirror)
+		if err != nil {
+			panic(err)
+		}
 
-	t.remote, err = remoteFromMirror(ctx, t.mirror)
-	if err != nil {
-		return nil, err
-	}
+		t.client = client.NewClient(t.local, t.remote)
 
-	t.client = client.NewClient(t.local, t.remote)
+		trustedMeta, err := t.local.GetMeta()
+		if err != nil {
+			panic(fmt.Errorf("getting trusted meta: %w", err))
+		}
 
-	trustedMeta, err := t.local.GetMeta()
-	if err != nil {
-		return nil, fmt.Errorf("getting trusted meta: %w", err)
-	}
+		// If the caller does not supply a root, then either use the root in the local store
+		// or default to the embedded one.
+		if root == nil {
+			root, err = getRoot(trustedMeta, t.embedded)
+			if err != nil {
+				panic(fmt.Errorf("getting trusted root: %w", err))
+			}
+		}
 
-	// If the caller does not supply a root, then either use the root in the local store
-	// or default to the embedded one.
-	if root == nil {
-		root, err = getRoot(trustedMeta, t.embedded)
-		if err != nil {
-			return nil, fmt.Errorf("getting trusted root: %w", err)
+		if err := t.client.InitLocal(root); err != nil {
+			panic(fmt.Errorf("unable to initialize client, local cache may be corrupt: %w", err))
 		}
-	}
 
-	if err := t.client.InitLocal(root); err != nil {
-		return nil, fmt.Errorf("unable to initialize client, local cache may be corrupt: %w", err)
-	}
+		// We may already have an up-to-date local store! Check to see if it needs to be updated.
+		trustedTimestamp, ok := trustedMeta["timestamp.json"]
+		if ok && !isExpiredTimestamp(trustedTimestamp) && !forceUpdate {
+			// We're golden so stash the TUF object for later use
+			singletonTUF = t
+			return
```

Comment:
Is this returning nil?

Reply:
It's returning out of the `sync.Once` func, but we still return `singletonTUF` from `initializeTUF`